### PR TITLE
Add new endpoint to decode missing transactions

### PR DESCRIPTION
**Pull request by yabirgb**
Allow to decode all transactions that are missing the decoding process

### CODE REVIEW

File: rotkehlchen/db/ethtx.py [hunk 1]
```diff
@@ -259,19 +264,26 @@ def get_transaction_hashes_no_receipt(
...
+        Get the transactions hashes for the transactions that have not been decoded. If the limit
+        argument is provided then it is used in the SQL query with the default order.
+        """
         cursor = self.db.conn.cursor()
-        querystr = (
-            'SELECT A.tx_hash from ethtx_receipts AS A LEFT OUTER JOIN evm_tx_mappings AS B '
-            'ON A.tx_hash=B.tx_hash WHERE B.tx_hash is NULL'
-        )
-        bindings = ()
+        querystr = 'SELECT A.tx_hash from ' + TRANSACTIONS_MISSING_DECODING_QUERY
```
Comment (LefterisJP) [inline rotkehlchen/db/ethtx.py:272]: Since you are doing this put the cursor also under the context manager

File: rotkehlchen/api/v1/resources.py [hunk 2]
```diff
@@ -467,6 +467,7 @@ def get(self) -> Response:
 class EthereumTransactionsResource(BaseMethodView):
     get_schema = EthereumTransactionQuerySchema()
     post_schema = EthereumTransactionDecodingSchema()
+    put_schema = AsyncQueryArgumentSchema()
```
Comment (LefterisJP) [inline rotkehlchen/api/v1/resources.py:470]: I am not sure if this is the best approach. Semantically this sounds wrong.

You are doing a `PUT /blockchains/ETH/transactions`. This is hard to imagine as being a request to decode any undecoded ethereum transactions. I would expect this to add a transaction.

I would prefer a new endpoint for this `/blockchains/ETH/transactions/decode`

File: rotkehlchen/tests/utils/ethereum.py [hunk 3]
```diff
@@ -277,9 +278,11 @@ def setup_ethereum_transactions_test(
         ],
     )
 
-    if one_receipt_in_db:
+    if one_receipt_in_db is True:
         with database.user_write() as cursor:
             dbethtx.add_receipt_data(cursor, txreceipt_to_data(expected_receipt1))
+            if second_receipt_in_db is True:
```
Comment (LefterisJP) [inline rotkehlchen/tests/utils/ethereum.py:284]: Why this change here?

**Review by LefterisJP — COMMENTED** [review]
Left comments. You also need to edit the api docs.

File: rotkehlchen/tests/utils/ethereum.py (hunk 3 above)
Comment (yabirgb) [reply rotkehlchen/tests/utils/ethereum.py:284]: because the previous implementation didn't add the second receipt in order to query it from etherscan and test that functionality. For this test I don't need to query etherscan as it is only checking that the decoding works as expected and wanted to reuse this logic that we already had

File: rotkehlchen/db/ethtx.py (hunk 1 above)
Comment (yabirgb) [reply rotkehlchen/db/ethtx.py:272]: I thought that and forgot. Will do

File: docs/api.rst [hunk 4]
```diff
@@ -1896,6 +1896,42 @@ Purging locally saved ethereum transactions
...
+
+
+   **Example Response**:
+
+   .. sourcecode:: http
+
+      HTTP/1.1 200 OK
+      Content-Type: application/json
+
+      { "result": {"amount_decoded": 4}, "message": "" }
```
Comment (LefterisJP) [inline docs/api.rst:1927]: Here and anywhere else you use it, please use `decoded_tx_number`.

Study the difference between `number` and `amount` in the english language to understand. It's a bit subtle.
https://www.enago.com/academy/understanding-the-difference-between-number-and-amount/#:~:text=What%20is%20the%20difference%2C%20then,things%20that%20cannot%20be%20counted.

```suggestion
      { "result": {"decoded_tx_number": 4}, "message": "" }
```

File: rotkehlchen/db/ethtx.py [hunk 5]
```diff
@@ -259,18 +264,25 @@ def get_transaction_hashes_no_receipt(
...
             querystr += ' LIMIT ?'
 
-        cursor.execute(querystr, bindings)
-        return [make_evm_tx_hash(x[0]) for x in cursor]
+        with self.db.conn.read_ctx() as cursor:
+            cursor.execute(querystr, bindings)
+            return [make_evm_tx_hash(x[0]) for x in cursor]
+
+    def count_hashes_not_decoded(self) -> int:
+        """Count the amount of transactions queried that have not been decoded"""
```
Comment (LefterisJP) [inline rotkehlchen/db/ethtx.py:282]:
```suggestion
        """Count the number of transactions queried that have not been decoded"""
```

**Review by LefterisJP — COMMENTED** [review]
some nitpicks

File: docs/api.rst [hunk 6]
```diff
@@ -1896,6 +1896,42 @@ Purging locally saved ethereum transactions
...
+   **Example Response**:
+
+   .. sourcecode:: http
+
+      HTTP/1.1 200 OK
+      Content-Type: application/json
+
+      { "result": {"decoded_tx_number": 4}, "message": "" }
+
+   :resjson integer decoded_tx_number: The amount of transactions that have been decoded.
```
Comment (LefterisJP) [inline docs/api.rst:1929]: Do I really have to start going through every single instance of what i need you guys to change and point it out? :disappointed: 

```suggestion
   :resjson integer decoded_tx_number: The number of transactions that have been decoded.
```

File: rotkehlchen/api/rest.py [hunk 7]
```diff
@@ -3321,6 +3321,30 @@ def decode_ethereum_transactions(
         result_dict = _wrap_in_result(result, msg)
         return api_response(process_result(result_dict), status_code=status_code)
 
+    def _decode_pending_ethereum_transactions(self) -> Dict[str, Any]:
+        dbethtx = DBEthTx(self.rotkehlchen.data.db)
+        amount_of_tx_to_decode = dbethtx.count_hashes_not_decoded()
```
Comment (LefterisJP) [inline rotkehlchen/api/rest.py:3326]: I guess I do ...

```suggestion
        number_of_tx_to_decode = dbethtx.count_hashes_not_decoded()
```

File: rotkehlchen/api/rest.py [hunk 8]
```diff
@@ -3321,6 +3321,30 @@ def decode_ethereum_transactions(
...
+        if amount_of_tx_to_decode > 0:
```
Comment (LefterisJP) [inline rotkehlchen/api/rest.py:3327]:
```suggestion
        if number_of_tx_to_decode > 0:
```

File: rotkehlchen/api/rest.py [hunk 9]
```diff
@@ -3321,6 +3321,30 @@ def decode_ethereum_transactions(
...
+            self.rotkehlchen.eth_tx_decoder.get_and_decode_undecoded_transactions()
+
+        return {
+            'result': {'decoded_tx_number': amount_of_tx_to_decode},
```
Comment (LefterisJP) [inline rotkehlchen/api/rest.py:3331]:
```suggestion
            'result': {'decoded_tx_number': number_of_tx_to_decode},
```

**Review by LefterisJP — APPROVED** [review]
I will merge but pretty dissapointed that in 4 places you still did not do what I asked. Note down and fix.
